File: app/src/main/java/dev/bee/kanjianki/MainActivityStudyWritingUi.java
package dev.bee.kanjianki;

import android.view.View;

import dev.bee.kanjianki.core.StudyTaskCopy;
import dev.bee.kanjianki.core.study.StrokeGuide;
import dev.bee.kanjianki.core.study.WritingActionPresentation;

final class MainActivityStudyWritingUi {
    private final MainActivityStudy activity;
    private final MainActivityStudyWritingStatus writingStatus;
    private final MainActivityStudyWritingToolbar writingToolbar;

    MainActivityStudyWritingUi(MainActivityStudy activity) {
        this.activity = activity;
        this.writingStatus = new MainActivityStudyWritingStatus(activity);
        this.writingToolbar = new MainActivityStudyWritingToolbar(activity);
    }

    void buildStudyActionBar() {
        writingToolbar.buildStudyActionBar();
    }

    void updateResultActions() {
        WritingActionPresentation presentation = writingActionPresentation();
        updateToolActionRow(presentation);
        updatePrimaryActionRow(presentation);
        updateFallbackActionButtons(presentation);
        updateHintAndAnswerVisibility(presentation);
        if (!presentation.resultStatusVisible) {
            hideResultStatus();
        }
    }

    void refreshWritingModelStatus() {
        writingStatus.refreshWritingModelStatus();
    }

    WritingActionPresentation writingActionPresentation() {
        WritingActionPresentation.Input input = new WritingActionPresentation.Input(activity.activeAnalysis);
        input.checkingWriting = activity.checkingWriting;
        input.canUndoStroke = activity.drawingPad != null && activity.drawingPad.canUndoStroke();
        input.writingModelStatusKnown = activity.writingModelStatusKnown;
        input.writingModelDownloaded = activity.writingModelDownloaded;
        input.hasReplaySnapshot = activity.drawingPad != null && activity.drawingPad.hasReplaySnapshot();
        input.hasInk = activity.drawingPad != null && activity.drawingPad.hasInk();
        input.guide = activity.activeSession == null ? null : activity.strokeGuide(activity.activeSession.item.kanji);
        input.canRevealMoreHelp = canRevealMoreHelp();
        input.recallTask = activity.activeSession != null && StudyTaskCopy.isRecallTask(activity.activeSession);
        input.teachingTask = activity.activeSession != null && StudyTaskCopy.isTeachingTask(activity.activeSession);
        input.currentPracticeLevel = activity.currentPracticeLevel;
        return WritingActionPresentation.from(input);
    }

    void updateUndoStrokeButton() {
        updateToolActionRow(writingActionPresentation());
    }

    void updateToolActionRow(WritingActionPresentation presentation) {
        if (activity.writingToolActionsView != null) {
            activity.writingToolActionsView.render(new WritingToolActionsModel(
                    presentation.undoEnabled,
                    presentation.hintText,
                    presentation.hintVisible,
                    activity::eraseWritingPad,
                    activity::undoWritingStroke,
                    activity::showWritingHint
            ));
        }
    }

    void updatePrimaryActionRow(WritingActionPresentation presentation) {
        if (activity.writingPrimaryActionsView != null) {
            activity.writingPrimaryActionsView.render(new WritingPrimaryActionsModel(
                    presentation.checkText,
                    presentation.checkVisible,
                    presentation.checkEnabled,
                    "Download checker",
                    presentation.downloadVisible,
                    presentation.nextLabel,
                    presentation.nextVisible,
                    presentation.messyPass ? activity::startCleanerRetry : activity::checkWriting,
                    writingStatus::downloadWritingModel,
                    () -> activity.submitReview(presentation.nextRating, false)
            ));
        }
    }

    void updateFallbackActionButtons(WritingActionPresentation presentation) {
        if (activity.writingFallbackActionsView != null) {
            activity.writingFallbackActionsView.render(new WritingFallbackActionsModel(
                    presentation.replayVisible,
                    presentation.manualOverrideVisible,
                    presentation.practiceWithGuideVisible,
                    activity::replayWritingAnalysis,
                    () -> activity.submitReview(activity.RATING_GOOD, true),
                    activity::startGuidedWritingRetry
            ));
        }
    }

    void updateHintAndAnswerVisibility(WritingActionPresentation presentation) {
        if (activity.writingAnswerPanelState != null) {
            activity.writingAnswerPanelState.updateVisible(presentation.answerPanelVisible);
        } else if (activity.studyAnswerPanel != null) {
            activity.studyAnswerPanel.setVisibility(presentation.answerPanelVisible ? View.VISIBLE : View.GONE);
        }
    }

    boolean canRevealMoreHelp() {
        if (activity.activeSession == null) {
            return false;
        }
        StrokeGuide guide = activity.strokeGuide(activity.activeSession.item.kanji);
        return activity.hintProgression.canRevealMoreHelp(activity.currentHintState, guide);
    }

    void setStudyStatus(String value, int color) {
        if (activity.studyStatus != null) {
            activity.studyStatus.setStatus(value, color);
        }
        if (activity.activeAnalysis == null) {
            hideResultStatus();
        }
    }

    void setResultStatus(String value, int color) {
        if (activity.writingResultStatus != null) {
            activity.writingResultStatus.show(value, color);
        }
    }

    void hideResultStatus() {
        if (activity.writingResultStatus != null) {
            activity.writingResultStatus.hide();
        }
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/MainActivityStudyWritingUi.kt
package dev.bee.kanjianki

import android.view.View
import dev.bee.kanjianki.core.StudyTaskCopy
import dev.bee.kanjianki.core.StudyWritingCopy
import dev.bee.kanjianki.core.study.WritingActionPresentation
import dev.bee.kanjianki.core.study.WritingFeedbackCopy

internal class MainActivityStudyWritingUi(private val activity: MainActivityStudy) {
    private val writingStatus = MainActivityStudyWritingStatus(activity)
    private val writingToolbar = MainActivityStudyWritingToolbar(activity)

    fun buildComposeActionBarState(): WritingActionsBarState {
        return writingToolbar.buildComposeActionBarState()
    }

    fun updateResultActions() {
        val presentation = writingActionPresentation()
        updateToolActionRow(presentation)
        updatePrimaryActionRow(presentation)
        updateFallbackActionButtons(presentation)
        updateHintAndAnswerVisibility(presentation)
        if (!presentation.resultStatusVisible) {
            hideResultStatus()
        }
    }

    fun refreshWritingModelStatus() {
        writingStatus.refreshWritingModelStatus()
    }

    fun writingActionPresentation(): WritingActionPresentation {
        val input = WritingActionPresentation.Input(activity.activeAnalysis)
        val drawingPad = activity.drawingPad
        val session = activity.activeSession
        input.checkingWriting = activity.checkingWriting
        input.canUndoStroke = drawingPad != null && drawingPad.canUndoStroke()
        input.writingModelStatusKnown = activity.writingModelStatusKnown
        input.writingModelDownloaded = activity.writingModelDownloaded
        input.hasReplaySnapshot = drawingPad != null && drawingPad.hasReplaySnapshot()
        input.hasInk = drawingPad != null && drawingPad.hasInk()
        input.guide = session?.item?.let { activity.strokeGuide(it.kanji) }
        input.canRevealMoreHelp = canRevealMoreHelp()
        input.repairTask = session != null && StudyTaskCopy.isRepairWritingTask(session)
        input.recallTask = session != null && StudyTaskCopy.isRecallTask(session)
        input.teachingTask = session != null && StudyTaskCopy.isTeachingTask(session)
        input.currentPracticeLevel = activity.currentPracticeLevel
        return WritingActionPresentation.from(input)
    }

    fun updateUndoStrokeButton() {
        updateToolActionRow(writingActionPresentation())
    }

    fun updateToolActionRow(presentation: WritingActionPresentation) {
        activity.writingToolActionsView?.render(
            WritingToolActionsModel(
                presentation.undoEnabled,
                presentation.hintText,
                presentation.hintVisible,
                Runnable { activity.eraseWritingPad() },
                Runnable { activity.undoWritingStroke() },
                Runnable { activity.showWritingHint() }
            )
        )
    }

    fun updatePrimaryActionRow(presentation: WritingActionPresentation) {
        val repairSkipVisible = activity.activeSimilarWritingRepair != null && !presentation.hasResult
        activity.writingPrimaryActionsView?.render(
            WritingPrimaryActionsModel(
                checkText = presentation.checkText,
                checkVisible = presentation.checkVisible,
                checkEnabled = presentation.checkEnabled,
                downloadText = WritingFeedbackCopy.downloadCheckerLabel(),
                downloadVisible = presentation.downloadVisible,
                nextText = if (repairSkipVisible) StudyWritingCopy.continueAnywayLabel() else presentation.nextLabel,
                nextVisible = presentation.nextVisible || repairSkipVisible,
                nextEnabled = if (repairSkipVisible) presentation.checkEnabled else true,
                onCheck = if (presentation.messyPass) {
                    Runnable { activity.startCleanerRetry() }
                } else {
                    Runnable { activity.checkWriting() }
                },
                onDownload = Runnable { writingStatus.downloadWritingModel() },
                onNext = if (repairSkipVisible) {
                    Runnable { activity.skipSimilarWritingRepair() }
                } else {
                    Runnable { activity.submitReview(presentation.nextRating, false) }
                },
                skipText = StudyWritingCopy.skipLabel(),
                skipVisible = presentation.skipVisible,
                skipEnabled = presentation.skipEnabled,
                onSkip = Runnable { activity.submitReview(MainActivityBase.RATING_GOOD, true) },

            )
        )
    }

    fun updateFallbackActionButtons(presentation: WritingActionPresentation) {
        activity.writingFallbackActionsView?.render(
            WritingFallbackActionsModel(
                presentation.replayVisible,
                presentation.manualOverrideVisible,
                presentation.practiceWithGuideVisible,
                Runnable { activity.replayWritingAnalysis() },
                Runnable { activity.submitReview(MainActivityBase.RATING_GOOD, true) },
                Runnable { activity.startGuidedWritingRetry() },
            )
        )
    }

    fun updateHintAndAnswerVisibility(presentation: WritingActionPresentation) {
        val answerPanelState = activity.writingAnswerPanelState
        if (answerPanelState != null) {
            answerPanelState.updateVisible(presentation.answerPanelVisible)
        } else {
            val answerPanel = activity.studyAnswerPanel
            if (answerPanel != null) {
                answerPanel.visibility = if (presentation.answerPanelVisible) View.VISIBLE else View.GONE
            }
        }
    }

    fun canRevealMoreHelp(): Boolean {
        val session = activity.activeSession ?: return false
        val guide = activity.strokeGuide(session.item?.kanji ?: return false)
        return activity.hintProgression.canRevealMoreHelp(activity.currentHintState, guide)
    }

    fun setStudyStatus(value: String, color: Int) {
        activity.studyStatus?.setStatus(value, color)
        if (activity.activeAnalysis == null) {
            hideResultStatus()
        }
    }

    fun setResultStatus(value: String, color: Int) {
        activity.writingResultStatus?.show(value, color)
    }

    fun hideResultStatus() {
        activity.writingResultStatus?.hide()
    }
}
